build/xtask/src/dist.rs

```diff
@@ -836,12 +949,80 @@ fn update_image_header(
                     ..Default::default()
                 };
 
-                for (i, (_, range)) in map.iter().enumerate() {
-                    header.sau_entries[i].rbar = range.start;
-                    header.sau_entries[i].rlar = (range.end - 1) & !0x1f | 1;
-                }
+                let last = if let Some(s) = secure {
+                    let mut i = 0;
+
+                    // Our memory layout with a secure task looks like the
+                    // following:
+                    // +---------------+
+                    // |               |
+                    // |   Task        |
+                    // | (Non-secure)  |
+                    // |               |
+                    // |               |
+                    // +---------------+
+                    // |               |
+                    // |   Task        |
+                    // | (Non-secure)  |
+                    // |               |
+                    // |               |
+                    // +---------------+
+                    // |               |
+                    // |   Task        |
+                    // | (Secure)      |
+                    // +---------------+
+                    // |    NSC        |
+                    // +---------------+
+                    // |               |
+                    // |   Task        |
+                    // | (Non-secure)  |
+                    // |               |
+                    // |               |
+                    // +---------------+
+                    //
+                    // The entries in the SAU specify regions that are
+                    // non-secure OR non-secure callable (NSC).
+                    // This means the entry for our flash gets broken
+                    // down into three entries:
+                    // 1) Non-secure range before the secure task
+                    // 2) non-secure range after the secure task
+                    // 3) NSC region in the secure task
+                    for (_, range) in map.iter() {
+                        if range.contains(&s.secure.start) {
+                            header.sau_entries[i].rbar = range.start;
+                            header.sau_entries[i].rlar =
+                                (s.secure.start - 1) & !0x1f | 1;
```

Comment:
comment or function to make it clearer what this line is doing.  I'm pretty confident this is setting 1 to mark the entry valid but it'd help to have it spelled out here and elsewhere.